### ISSUE TITLE
Rotation snapping is active only when it is turned off.

### ISSUE THREAD

**Issue by avencherus** [godotengine/godot#23072]
**Win10 64bit - Alpha 3.1 0dbe01483a902c49ecedf4fd36b74353424145a5**

@groud Likely some flags flipped up somewhere.  But have noticed that rotation snapping only works when both snapping options are disabled.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/13004169/47065614-4b437800-d1ec-11e8-9068-19adf6fd1b37.gif)

**Comment by groud** [comment]
You have to release the ctrl key while dragging, it switches the current snapping state

**Comment by avencherus** [comment]
Oh thanks.  I see.  That was unexpected.  XD

It is still a little bit confusing, since the behavior flips when the options are changed.

So when the snapping is on, the CTRL drag does no snapping until CTRL is released.  There is no immediate indicators that the mode is being changed and won't release until CTRL is released.

I'm not sure how many people have the same habit as me, but yep, I'm holding that CTRL button because I expect bad things will happen if I let go in the middle.  XD

Using it now seems like it can be very handy and time saving, but no ideas on how to make it more clear without some sort of universal way of toggling snapping with modifier keys.

Maybe flipping the behavior is still a way to go, and adding tooltip information?  That way if one doesn't know or care, they would at least start with the snapping behavior that the options reflect.  Then if they let go, they would then discover this alternative mode, or maybe find it the tooltips.

**Comment by groud** [comment]
I think I should simply switch the snapping icon status when control is pressed.
I don't think flipping the behavior is ok, people with complain in any of the two situation.

Using other kind of shortcuts (like blender does with g, s and all) is probably the way to go at some point.

**Comment by akien-mga** [comment]
Is this still reproducible in the current master branch?

**Comment by avencherus** [comment]
Yes.

Currently, when rotation snapping is **off**, a CTRL drag will initially do snapping rotation, until CTRL is released.  Then the non-snapping rotation kicks in.

Then when rotation snapping is **on**, a CTRL drag will first do a non-snapping rotation, until CTRL is released.

I wouldn't contest the releasing CTRL to toggle functionality itself, that is useful and working.  But I do believe the flagging is inverted and catches one off guard.

**Comment by Anutrix** [comment]
Why not flip the snap/magnet icon too when control is held? Just a visibility change and it won't even confuse people.

**Comment by KoBeWi** [comment]
Toggling the snap icon with control is easy to do, but I noticed that snap options menu has "Use Rotation Steps" and "Use Scale Snap" options, which work completely independently from the snap button (and scale snapping isn't affected by the standard snap button at all) and the CTRL will always disable them temporarily when they are enabled.

So the new problem is what to do about these options and why they even exist. I assume unifying the snapping to one button (which is now even more confusingly called "Use Smart Snapping", even though it affects rotation) would be the correct solution. Then making the Ctrl toggle this button on pressed/released.

**Comment by avencherus** [comment]
@KoBeWi I agree, it's confusing.  I imagine  smart snapping to be something that tries to align with relevant points nearby, and the normal kind would force things to very specific steps in units or degrees.

I don't think they can sensibly combine, the target of smart snapping might be things that aren't snapped to the desired grid.

Does smart snap deal with alignment of rotations?

If they are too be exclusive, then maybe they are provided with a radio button.

**Comment by KoBeWi** [comment]
Still valid in 3.2.4 beta3 (see my last comment)

**Comment by foxydevloper** [comment]
Still an issue in v4.0.dev.custom_build [d17201afc]
Inside of 2D:
- If "Use Rotation Snap" is off, doing Ctrl+Drag with the select tool rotates with snapping. Letting go of Ctrl stops the snapping.
- If you enable "Use Rotation Snap", whether Ctrl causes snapping is inverted.

Inside of 3D, Ctrl+Drag with the select tool also rotates with snapping, however letting go of ctrl does not stop it from snapping.

"Use Rotation Snap" inverting whether or not Ctrl causes snapping is most likely intended behavior. The issue is the fact that Ctrl+Drag  uses Ctrl, which conflicts with the shortcut for snapping. So every time you Ctrl+Drag to rotate it causes snapping, and if you have "Use Rotation Snap" enabled, since you're pressing Ctrl it stops snapping from occurring.
